ambari-server/src/main/java/org/apache/ambari/server/controller/internal/ClusterTemplateArtifactPasswordReplacer.java

```diff
@@ -0,0 +1,138 @@
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+
+import java.util.AbstractMap;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import org.apache.ambari.server.state.StackId;
+import org.apache.ambari.server.topology.Configurable;
+import org.apache.ambari.server.topology.Configuration;
+import org.apache.ambari.server.utils.SecretReference;
+
+import com.google.common.collect.Multimap;
+
+/**
+ * Helper class for replacing password properties in cluster template artifacts. {@see #replacePasswords}
+ */
+public class ClusterTemplateArtifactPasswordReplacer {
+
+  /**
+   * Replaces all passwords (service config passwords and default password) in a received cluster creation template
+   * artifact. Mpack (stack) information is used to identify password type properties. If the cluster template does not
+   * specify mpacks, all installed stacks are queried for password type properties.
+   *
+   * @param artifactData the raw cluster template artifact as parsed json
+   * @return the cluster template artifact with passwords replaced
+   */
+  public Map<String, Object> replacePasswords(Map<String, Object> artifactData) {
+    Collection<StackId> stackIds = extractStackIdsFromClusterRequest(artifactData);
+    // get all password properties from the specified stacks or
+    // all stacks if the cluster template doesn't specify mpacks
+    Multimap<String, String> passwordProperties = stackIds.isEmpty() ?
+      SecretReference.getAllPasswordProperties() :
+      SecretReference.getAllPasswordProperties(stackIds);
+    Map<String, Object> passwordsReplaced = replacePasswordsInConfigurations(artifactData, passwordProperties);
+    passwordsReplaced.replace("default_password", SecretReference.SECRET_PREFIX + ":default_password");
+    return passwordsReplaced;
+  }
+
+  /**
+   * Replaces passwords in the received cluster template artifact based on the received password information extracted
+   * from stacks.
+   * @param artifactData the cluster template artifact
+   * @param passwordProperties a multimap containing password type properties. The map has a structure of
+   *                           config type -> password properties.
+   * @return the cluster template artifact with passwords replaced
+   */
+  protected Map<String, Object> replacePasswordsInConfigurations(Map<String, Object> artifactData,
+                                                                 Multimap<String, String> passwordProperties) {
+    return (Map<String, Object>)applyToAllConfigurations(artifactData,
+      config -> {
+        Configuration configuration = Configurable.parseConfigs(config);
+        Configuration replacedConfiguration =
+          SecretReference.replacePasswordsInConfigurations(configuration, passwordProperties);
+        return Configurable.convertConfigToMap(replacedConfiguration);
+      }
```

Comment:
Can you please extract this lambda expression to a function that can be referenced and tested more easily?

Reply:
This method simply transforms a map with a lambda that is only 3 statements. I don't think further decomposition should be needed.